examples/ipsec-inline.p4

```diff
@@ -0,0 +1,376 @@
+                             in bit<256> key,
+                             in bit<9> key_size,
+                             in bit<1> ext_esn_en,
+                             in bit<1> enable_auth,
+                             in bit<64> esn) {
+
+        // Initialize the ipsec accelerator
+        // esn = esn + 1;
+
+        // Set IV information needed for encryption
+        // For ipsec combine salt and esn
+        bit<128> iv = (bit<128>)(salt ++ esn);
+        ipsec_acc.set_iv(iv);
+
+        ipsec_acc.set_key(key, key_size);
+
+        // For tunnel mode, operation, copy original IP header that needs to
+        // be encrypted. This header will be emitted after ESP header.
+        // Add protocol specific headers to the packet (rfc4106)
+        // 32bit seq number is used
+        // update tunnel ip header
+        // Set outer header's next header as ESP
+        // payload_offset : points inner(original) ip header which follows the esp header
+        // Done by the extern inline
+        ipsec_acc.set_auth_data_offset(0);
+        ipsec_acc.set_auth_data_len(0);
+        ipsec_acc.set_payload_offset(0);
+        ipsec_acc.set_payload_len(0);
+
+        // TODO: compute padding, build esp_trailer etc.
+
+        // instruct engine to add icv after encrypted payload
+        ipsec_acc.set_icv_offset(ICV_AFTER_PAYLOAD);
+        ipsec_acc.set_icv_len(4); // Four bytes of ICV value.
+
+        // run encryption w/ authentication
+        if (ipsec_acc.encrypt_inline(crypto_mode_e.TUNNEL, 
+                                 enable_auth, 
+                                 spi, 
+                                 esn[31:0], 
+                                 iv) !=
+            crypto_results_e.SUCCESS) {
+            // TODO:
+            // Check if this is AUTH error or some other error.
+            // Drop the packet or do other things as needed
+            drop_packet();
+            return;
+        }
+    }
+
+    action ipsec_sa_action(bit<32>    spi,
+                           bit<32>    salt,
+                           bit<256>   key,
+                           bit<9>     key_size,
+                           bit<1>     ext_esn_en,
+                           bit<1>     auth_en,
+                           bit<1>     valid_flag,
+                           bit<64>    esn) {
+        if (valid_flag == 0) {
+            return;
+        }
+        if (!hdr.esp.isValid()) {
+            ipsec_esp_encrypt(spi, salt, key, key_size, ext_esn_en, auth_en, esn);
+        } else {
+            ipsec_esp_decrypt(spi, salt, key, key_size, ext_esn_en, auth_en, esn);
+        }
+    }
+    // setup crypto accelerator for decryption - get info from sa table
+    // lookup sa_table using esp.spi
+    table ipsec_sa {
+        key = {
+            // For encrypt case get sa_idx from parser
+            // for decrypt case esp hdr spi value will be used as sa_idx
+            main_meta.sa_index  : exact;
```

Comment:
Would it make sense to have separate match-action tables for encryption and decryption?

For example, encrypt table could match on the destination IP address, while decrypt table could match on source and destination IP address, as well as SPI value. (This is similar to the approach proposed in [P4-IPsec](https://ieeexplore.ieee.org/document/9151942))

Reply:
Sure, that's possible. This is just an example, to focus on how the inline model works

Reply:
Actually, what you proposed is simpler than setting parser metadata. I'll look at rewriting the example app.